src/ripple/app/tx/impl/CancelTicket.cpp

```diff
@@ -80,15 +80,18 @@ CancelTicket::doApply ()
 
     std::uint64_t const hint (sleTicket->getFieldU64 (sfOwnerNode));
 
-    auto viewJ = ctx_.app.journal ("View");
-    TER const result = dirDelete (ctx_.view (), false, hint,
-        keylet::ownerDir (ticket_owner), ticketId, false, (hint == 0), viewJ);
+    if (! ctx_.view().dirRemove(
+            keylet::ownerDir(ticket_owner), hint, ticketId, false))
+    {
+        return tefBAD_LEDGER;
+    }
 
+    auto viewJ = ctx_.app.journal ("View");
     adjustOwnerCount(view(), view().peek(
```

Comment:
Good spotting, @mellery451.  This one is messy.  I'll give you my impression of what is afoot here.  We're dealing with a case where a `dirRemove()`, of an object that had better be in the directory, failed.  The failing case _should_ never happen.  If it were to happen it's not clear what the best recovery position would be. And the failing case is hard enough to reproduce that there are no unit tests for it.

So what do we know?

1. If `dirDelete()` had failed in this way, then it would return `tefBAD_LEDGER`.  So we're matching old behavior there.  That's (probably) good.
2. If `doApply()` (the method we're in) returns a code other than `tesSUCCESS` then the transaction is not applied.  So all of the changes we're making here will be rolled back.  (That's _almost_ true -- I'm eliding the `tec`code case since it's not relevant here.)
3. Therefore the new code _should_ accomplish the same ends as the old code but slightly more efficiently.  Since the change to `ownerCount` and the erasure of the `Ticket` will be rolled back because we're returning a `tef` code we don't _need_ to perform those actions.  But there's no harm if we did, since they'll be rolled back.
4. However the extra efficiency is probably irrelevant, since we're hoping this error condition will never occur.  If it does, it means something got fouled up in the ledger.  There should be no way that a _user_ could cause this error to occur.

_To err is human.  To really foul things up requires a computer_ -- Paul Ehrlich

In summary, I _think_ the code is fine as it stands.  And to me it's a hair easier to read.  But it's also fine to switch it back to the previous formulation.

Reply:
sounds good - thanks for the explanation.